### PR TITLE
Feature/object validation

### PR DESCRIPTION
**Pull request by bfolie**
Square validation and (lack of) cycle validation

### CODE REVIEW

File: docs/faq.md [hunk 1]
```diff
@@ -33,3 +33,10 @@ If there are multiple properties and/or multiple statistics, they should all go
 The statistics should be
 [`Property`](../specification/attributes/#property)
 attributes with their `origin` field set to `computed`.
+
+## How do I represent repeated applications of the same process?
+
+Consider a situation in which a process repeatedly applied to a material, but does not substantially change the material.
```
Comment (bfolie) [inline docs/faq.md]:
```suggestion
Consider a situation in which a process is repeatedly applied to a material, but does not substantially change the material.
```

File: docs/faq.md [hunk 2]
```diff
@@ -33,3 +33,10 @@ If there are multiple properties and/or multiple statistics, they should all go
 The statistics should be
 [`Property`](../specification/attributes/#property)
 attributes with their `origin` field set to `computed`.
+
+## How do I represent repeated applications of the same process?
+
+Consider a situation in which a process is repeatedly applied to a material, but does not substantially change the material.
+For example, repeated application of heat-treatment to harden a ceramic.
```
Comment (kroenlein) [inline docs/faq.md]:
```suggestion
For example, repeated application of heat-treatment to harden a ceramic or multiple coats of paint.
```

File: docs/faq.md [hunk 3]
```diff
@@ -33,3 +33,10 @@ If there are multiple properties and/or multiple statistics, they should all go
...
+Because material histories are chronological, each application must be a new process and must produce a new output material.
+The materials/processes may reference the same templates, however.
```
Comment (kroenlein) [inline docs/faq.md]:
```suggestion
The materials/processes may reference the same templates, however; the same attributes are relevant for each iteration.
```

File: docs/philosophy.md [hunk 4]
```diff
@@ -98,3 +98,11 @@ which must be used when recording physical information into [Attributes](../spec
...
 However, uncertainty quantification can be challenging and is often impossible in hindsight.
 Therefore, `nominal` values are permitted to express "It was nominally X, but we don't know how certain we are about it."
+
+
+#### Material history is chronological
+
+Starting from a given material, one can trace back the process that created the material, 
+the ingredients that went into the process, the materials that those ingredients were derived from, and so on.
+This history is chronological: for example, the ingredients that are involved in a process must exist *before* the process occurs.
+As a consequence, there can be no cycles in a material history.
```
Comment (kroenlein) [inline docs/philosophy.md]: I can't figure out how to do multiline-suggestions, and spawning a PR seems a bit much.  I'd propose this reads:

For any physical object, one can trace back the series of processes that generated it and the ingredient materials that went into those processes.  
Formally, such a chronology represents a directed acyclic graph, where causal structure leads naturally to a topological ordering for nodes.  
Following this formalism and the common-sense idea that time moves in one direction only, a material history cannot contain a cycle, i.e., following references in the order material->process->ingredient->material will never see the same node twice.

File: docs/philosophy.md [hunk 5]
```diff
@@ -98,3 +98,11 @@ which must be used when recording physical information into [Attributes](../spec
 
 However, uncertainty quantification can be challenging and is often impossible in hindsight.
 Therefore, `nominal` values are permitted to express "It was nominally X, but we don't know how certain we are about it."
+
+
+#### Material history is chronological
+
+Starting from a given material, one can trace back the process that created the material,
```
Comment (bfolie) [inline docs/philosophy.md]:
```suggestion
For any given physical object, one can trace back the series of processes that generated it and the ingredient materials that went into those processes. 
```

File: docs/philosophy.md [hunk 6]
```diff
@@ -98,3 +98,11 @@ which must be used when recording physical information into [Attributes](../spec
 
 However, uncertainty quantification can be challenging and is often impossible in hindsight.
 Therefore, `nominal` values are permitted to express "It was nominally X, but we don't know how certain we are about it."
+
+
+#### Material history is chronological
+
+Starting from a given material, one can trace back the process that created the material, 
+the ingredients that went into the process, the materials that those ingredients were derived from, and so on.
```
Comment (bfolie) [inline docs/philosophy.md]:
```suggestion
```

File: docs/philosophy.md [hunk 7]
```diff
@@ -98,3 +98,11 @@ which must be used when recording physical information into [Attributes](../spec
...
+This history is chronological: for example, the ingredients that are involved in a process must exist *before* the process occurs.
```
Comment (bfolie) [inline docs/philosophy.md]:
```suggestion
We can represent this chronological history as a directed acyclic graph where each material links to the process that created it and each process links to the materials that went into it.
Each of those input materials must exist *before* the process occurs.
```

File: docs/philosophy.md (hunk 4 above)
Comment (bfolie) [inline docs/philosophy.md]:
```suggestion
As a consequence there can be no cycles in a well-formed material history, as that would imply a material being created and then traveling back in time to be used in one of the processes that were involved in creating it.
```

File: docs/philosophy.md [hunk 8]
```diff
@@ -98,3 +98,12 @@ which must be used when recording physical information into [Attributes](../spec
 
 However, uncertainty quantification can be challenging and is often impossible in hindsight.
 Therefore, `nominal` values are permitted to express "It was nominally X, but we don't know how certain we are about it."
+
+
+#### Material history is chronological
+
+For any given physical object, one can trace back the series of processes that generated it and the ingredient materials that went into those processes. 
+We can represent this chronological history as a directed acyclic graph where each material links to the process that created it and each process links to the materials that went into it.
```
Comment (bfolie) [inline docs/philosophy.md]:
```suggestion
We represent this chronological history as a directed acyclic graph where each material links to the process that created it and each process links to the materials that went into it.
```

**Pull request merged**